File: spikes/009-ai-ide-extensions/workspace/sample-code/test_completions.go
// Package main provides sample Go code for testing AI completions.
//
// Test scenarios:
// 1. Struct methods - implement interfaces
// 2. Error handling - custom errors, wrapping
// 3. Goroutines - channels, sync patterns
// 4. Testing - table-driven tests
package main

import (
	"context"
	"errors"
	"fmt"
	"sync"
)

// User represents a user in the system.
type User struct {
	ID     int64
	Name   string
	Email  string
	Active bool
}

// UserError represents user-related errors.
type UserError struct {
	Op  string
	ID  int64
	Err error
}

func (e *UserError) Error() string {
	return fmt.Sprintf("%s: user %d: %v", e.Op, e.ID, e.Err)
}

func (e *UserError) Unwrap() error {
	return e.Err
}

// Test: Ask AI to add Is() and As() methods

// UserRepository manages user storage.
type UserRepository struct {
	mu    sync.RWMutex
	users map[int64]*User
}

// NewUserRepository creates a new repository.
func NewUserRepository() *UserRepository {
	return &UserRepository{
		users: make(map[int64]*User),
	}
}

// Test: Ask AI to complete these methods with proper locking
func (r *UserRepository) Get(ctx context.Context, id int64) (*User, error) {
	// TODO: Implement with context cancellation
	return nil, errors.New("not implemented")
}

func (r *UserRepository) Insert(ctx context.Context, user *User) error {
	// TODO: Implement with proper locking
	return errors.New("not implemented")
}

// Test: Ask AI to add FindByEmail method
// Test: Ask AI to add FindActive method that uses goroutines

// UserService provides business logic for users.
type UserService interface {
	GetUser(ctx context.Context, id int64) (*User, error)
	CreateUser(ctx context.Context, name, email string) (*User, error)
	// Test: Ask AI to expand this interface
}

// Test: Ask AI to implement UserService interface

// ValidateEmail checks if an email address is valid.
func ValidateEmail(email string) error {
	// Let AI complete the validation logic
	return nil
}

// FetchUserData retrieves user data from an external API.
// Test: Ask AI to complete with proper HTTP handling and context
func FetchUserData(ctx context.Context, url string) (*User, error) {
	return nil, errors.New("not implemented")
}

// Test: Ask AI to generate table-driven tests for ValidateEmail
// Test: Ask AI to add OpenTelemetry tracing to UserRepository
// Test: Ask AI to implement a circuit breaker for FetchUserData

File: spikes/009-ai-ide-extensions/workspace/sample-code/test_completions.rs
//! Sample Rust file for testing AI code completions.
//!
//! Test scenarios:
//! 1. Struct implementation - derive traits, impl blocks
//! 2. Error handling - Result types, custom errors
//! 3. Iterator methods - map, filter, collect chains
//! 4. Async patterns - tokio, async/await

use std::collections::HashMap;

/// Represents a user in the system.
#[derive(Debug, Clone)]
pub struct User {
    pub id: u64,
    pub name: String,
    pub email: String,
    pub active: bool,
}

// Test: Ask AI to implement Display, PartialEq, Hash for User

/// Custom error type for user operations.
#[derive(Debug)]
pub enum UserError {
    NotFound(u64),
    InvalidEmail(String),
    // Test: Ask AI to add more variants
}

// Test: Ask AI to implement std::error::Error for UserError

/// User repository for managing users.
pub struct UserRepository {
    users: HashMap<u64, User>,
}

impl UserRepository {
    pub fn new() -> Self {
        Self {
            users: HashMap::new(),
        }
    }

    // Test: Ask AI to complete CRUD methods
    pub fn get(&self, id: u64) -> Result<&User, UserError> {
        todo!("Implement get")
    }

    pub fn insert(&mut self, user: User) -> Result<(), UserError> {
        todo!("Implement insert")
    }

    // Test: Ask AI to add find_by_email method
    // Test: Ask AI to add find_active_users method using iterators
}

// Test: Ask AI to complete this function with proper error handling
pub fn validate_email(email: &str) -> Result<(), UserError> {
    // Let AI implement email validation
    todo!()
}

// Test: Ask AI to convert this to async with tokio
pub fn fetch_user_data(url: &str) -> Result<User, Box<dyn std::error::Error>> {
    todo!("Implement HTTP fetch")
}

// Test: Ask AI to generate unit tests for UserRepository
// Test: Ask AI to add serde serialization to User
// Test: Ask AI to implement From<User> for UserDto
